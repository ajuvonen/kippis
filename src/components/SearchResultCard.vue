<script setup lang="ts">
import {storeToRefs} from 'pinia';
import {useI18n} from 'vue-i18n';
import type {SearchResultDrink} from '@/utils/types';
import {useCocktailStore} from '@/stores/cocktail';
import IconComponent from '@/components/IconComponent.vue';

defineProps<{
  item: SearchResultDrink;
}>();

const {t} = useI18n();

const cocktailStore = useCocktailStore();
const {addToSelection, removeFromSelection} = cocktailStore;
const {selection} = storeToRefs(cocktailStore);
</script>
<template>
  <div class="search-result__wrapper">
    <div class="search-result__image-wrapper" role="button" tabindex="0">
      <div
        :style="{background: `url(${item.thumb})`, backgroundSize: 'cover'}"
        class="search-result__image"
        role="presentation"
      ></div>
      <div class="search-result__shadow">
        {{ item.name }}
      </div>
    </div>
    <button
      v-if="!selection.has(item.id)"
      :aria-label="t('searchResults.addDrink', [item.name])"
      class="search-result__action-button"
      @click="addToSelection(item.id)"
    >
      <IconComponent icon="plus" />
    </button>
    <button
      v-else
      :aria-label="t('searchResults.removeDrink', [item.name])"
      class="search-result__action-button bg-red-300 border-slate-800"
      @click="removeFromSelection(item.id)"
    >
      <IconComponent icon="trashCan" />
    </button>
  </div>
</template>
<style lang="scss" scoped>
.search-result__wrapper {
  @apply relative shadow-md rounded-md hover:cursor-pointer hover:shadow-lg transition-shadow ease-in-out;
}

.search-result__image-wrapper {
  @apply relative overflow-hidden w-40 h-40 rounded-md;
  outline-offset: 0.2rem;
}

.search-result__image {
  @apply h-full;
  filter: contrast(115%);
  transition: transform 0.5s ease-in-out;
  &:hover {
    transform: scale(1.1) rotate(-5deg);
  }
}

.search-result__shadow {
  @apply absolute w-full h-full top-0 pointer-events-none bg-gradient-to-b from-slate-800 to-transparent to-40% text-white uppercase text-xs p-2;
}

.search-result__action-button {
  @apply absolute -bottom-4 right-2 p-0 rounded-full w-10 text-lg;
}
</style>
