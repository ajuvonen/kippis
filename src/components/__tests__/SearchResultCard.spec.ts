import {mount} from '@vue/test-utils';
import {describe, it, expect, vi} from 'vitest';
import SearchResultCard from '@/components/SearchResultCard.vue';
import {testSearchResults} from '@/components/__tests__/mswHandlers';

describe('SearchResultCard.vue', () => {
  const cocktail = testSearchResults[0][1];

  it('mounts', () => {
    const wrapper = mount(SearchResultCard, {
      props: {
        preview: true,
        cocktail,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('calls openCocktailModal when the image wrapper is clicked', async () => {
    const openCocktailModal = vi.fn();
    const wrapper = mount(SearchResultCard, {
      props: {
        preview: true,
        cocktail,
      },
      global: {
        mocks: {
          openCocktailModal,
        },
      },
    });

    await wrapper.find('.search-result__image-wrapper').trigger('click');
    expect(openCocktailModal).toHaveBeenCalledWith(0);
  });

  it('calls openCocktailModal when the image wrapper is focused and Enter is pressed', async () => {
    const openCocktailModal = vi.fn();
    const wrapper = mount(SearchResultCard, {
      props: {
        preview: true,
        cocktail,
      },
      global: {
        mocks: {
          openCocktailModal,
        },
      },
    });

    const imageWrapper = wrapper.find('.search-result__image-wrapper');
    await imageWrapper.trigger('focus');
    await imageWrapper.trigger('keypress.enter');
    expect(openCocktailModal).toHaveBeenCalledWith(0);
  });
});
