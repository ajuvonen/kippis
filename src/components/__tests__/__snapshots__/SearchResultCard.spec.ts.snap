// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchResultCard.vue > mounts 1`] = `
"<div data-v-cd13f5f7="" class="search-result">
  <div data-v-cd13f5f7="" aria-label="View info about &quot;Margarita&quot;" class="search-result__image-wrapper" role="button" tabindex="0">
    <div data-v-35e45732="" data-v-cd13f5f7="" class="search-result__image">
      <div data-v-35e45732="" class="lazy-image__fallback"><svg data-v-35e45732="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
          <path d="M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z" class="transition-all"></path>
        </svg></div>
    </div>
    <div data-v-cd13f5f7="" class="search-result__shadow">Margarita</div>
  </div><button data-v-634d56d3="" data-v-cd13f5f7="" aria-label="Add &quot;Margarita&quot; to Selection" class="add-cocktail-0 action-button"><svg data-v-634d56d3="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
      <path d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z" class="transition-all"></path>
    </svg></button>
</div>"
`;
